src/RoadRegistry.BackOffice.Api/ZipArchiveWriters/ForEditor/RoadSegmentsToZipArchiveWriter.cs

```diff
@@ -38,16 +47,41 @@ public async Task WriteAsync(ZipArchive archive, EditorContext context, Cancella
                 RoadSegmentDbaseRecord.Schema
             );
             await using (var dbfEntryStream = dbfEntry.Open())
-            using (var dbfWriter =
-                new DbaseBinaryWriter(
-                    dbfHeader,
-                    new BinaryWriter(dbfEntryStream, _encoding, true)))
+            using (var dbfWriter = new DbaseBinaryWriter(dbfHeader, new BinaryWriter(dbfEntryStream, _encoding, true)))
             {
-                var dbfRecord = new RoadSegmentDbaseRecord();
-                foreach (var data in context.RoadSegments.OrderBy(_ => _.Id).Select(_ => _.DbaseRecord))
+                foreach (var batch in context.RoadSegments
+                    .OrderBy(_ => _.Id)
+                    .Select(_ => _.DbaseRecord)
+                    .AsEnumerable()
+                    .Batch(_zipArchiveWriterOptions.RoadSegmentBatchSize))
                 {
-                    dbfRecord.FromBytes(data, _manager, _encoding);
-                    dbfWriter.Write(dbfRecord);
+                    var dbfRecords = batch
+                        .Select(x =>
+                        {
+                            var dbfRecord = new RoadSegmentDbaseRecord();
+                            dbfRecord.FromBytes(x, _manager, _encoding);
+                            return dbfRecord;
+                        })
+                        .ToList();
+
+                    var cachedStreetNameIds = dbfRecords
+                        .Select(record => record.LSTRNMID.Value)
+                        .Union(dbfRecords.Select(record => record.RSTRNMID.Value))
+                        .Where(streetNameId => streetNameId.HasValue)
+                        .Select(streetNameId => streetNameId.Value);
```

Comment:
Perhaps materialize (ToList / ToArray) and distinct these?

Reply:
Materialization helps with readable stack traces - it's going to be materialized by composing the query anyway.

Reply:
Union already make a distinct list out of this, no?

Reply:
Yup, had to look it up - seems it does.